File: packages/sdk-py/aegis/quickstart.py
"""One-call onboarding helper for the AEGIS Python SDK — Round 25 seed.

Mirror of the TS SDK's ``Aegis.quickstart()``. Collapses the canonical
5-step flow (generate keypair → register agent → mint policy → sign
per-request → verify) into a single call so juniors hit a working
``verify`` in <60 seconds.

Example:

    import asyncio
    from aegis.quickstart import quickstart

    async def main():
        bundle = await quickstart(label="my-first-agent")
        token = await bundle.sign(action="commerce.purchase", amount=100)
        async with bundle.client as aegis:
            result = await aegis.verify(token)
            print("ok" if result.valid else result.denial_reason)

    asyncio.run(main())

The keypair lives on disk at ``~/.aegis/keys/<label>.json`` (mode 0600)
when the filesystem is writable, in memory otherwise (Lambda).
"""
from __future__ import annotations

import os
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Any, Awaitable, Callable, Optional

from .client import AsyncAegis
from .crypto import (
    SignContext,
    generate_keypair,
    sign_agent_token,
)
from .key_storage import KeyStorage, StoredKey, default_key_storage
from .models import AgentRecord, PolicyRecord, PolicyScope
from .runtime import PythonRuntime, detect_runtime

DEFAULT_LABEL = "aegis-quickstart"
DEFAULT_RUNTIME = "anthropic"
DEFAULT_POLICY_TTL = timedelta(hours=24)
DEFAULT_SCOPES: list[dict[str, Any]] = [
    {
        "category": "commerce",
        "spendLimit": {
            "currency": "USD",
            "maxPerTransaction": 100,
            "maxPerDay": 500,
        },
    },
]


@dataclass
class QuickstartBundle:
    """Returned by :func:`quickstart`."""

    client: AsyncAegis
    """Configured AEGIS client (entered context manager on first await use)."""

    agent: AgentRecord
    """The registered agent."""

    policy: PolicyRecord
    """The freshly-minted policy."""

    sign: Callable[..., Awaitable[str]]
    """Pre-bound signer. Pass the same kwargs as :func:`sign_agent_token`'s
    ``SignContext`` (``action``, ``amount``, ``currency``, ``merchant_domain``,
    ``merchant_id``, ``ttl_seconds``). Returns the signed token string.
    """

    runtime: PythonRuntime
    """Detected runtime for telemetry / debugging."""

    storage: KeyStorage
    """The KeyStorage adapter that holds the keypair."""

    key_name: str
    """Storage key name, in case the caller wants to rebind later."""


def _resolve_api_key(explicit: Optional[str]) -> str:
    """Resolve API key from explicit arg, falling back to ``AEGIS_API_KEY`` env."""
    if explicit:
        return explicit
    from_env = os.environ.get("AEGIS_API_KEY", "")
    if from_env:
        return from_env
    raise ValueError(
        "quickstart: api_key not provided. "
        "Set AEGIS_API_KEY in your environment, or pass api_key= to quickstart() "
        "(https://docs.aegislabs.io/errors/auth_required)",
    )


async def quickstart(
    *,
    api_key: Optional[str] = None,
    principal_id: Optional[str] = None,
    label: str = DEFAULT_LABEL,
    runtime: str = DEFAULT_RUNTIME,
    storage: Optional[KeyStorage] = None,
    key_name: Optional[str] = None,
    scopes: Optional[list[dict[str, Any]] | list[PolicyScope]] = None,
    policy_expires_at: Optional[datetime] = None,
    base_url: Optional[str] = None,
) -> QuickstartBundle:
    """One-call onboarding. Returns a :class:`QuickstartBundle`.

    Args:
        api_key: AEGIS API key. Falls back to ``AEGIS_API_KEY`` env.
        principal_id: AEGIS principal id. Required for registration.
        label: Human-readable label for the agent. Default ``"aegis-quickstart"``.
        runtime: Agent runtime. Default ``"anthropic"``.
        storage: KeyStorage adapter. Default: runtime-appropriate (filesystem
            on writable runtimes, memory on Lambda).
        key_name: Storage key name. Defaults to ``label``.
        scopes: Override the default permissive policy scopes.
        policy_expires_at: Override the default 24-hour expiry.
        base_url: Override the AEGIS API endpoint.
    """
    resolved_key = _resolve_api_key(api_key)
    if not principal_id:
        principal_id = os.environ.get("AEGIS_PRINCIPAL_ID", "")
    if not principal_id:
        raise ValueError(
            "quickstart: principal_id not provided. "
            "Pass principal_id= or set AEGIS_PRINCIPAL_ID env var.",
        )

    name = key_name or label
    store = storage or default_key_storage()
    detected = detect_runtime()

    client_kwargs: dict[str, Any] = {"api_key": resolved_key}
    if base_url:
        client_kwargs["base_url"] = base_url
    client = AsyncAegis(**client_kwargs)

    # Step 1 — load or generate the keypair.
    stored = await store.get(name)
    if stored is None:
        kp = generate_keypair()
        stored = StoredKey(
            private_key=kp.private_key,
            public_key=kp.public_key,
            created_at=datetime.now(timezone.utc).isoformat(),
            label=label,
        )
        await store.put(name, stored)

    # Step 2 — register or reuse.
    if stored.agent_id:
        try:
            agent = await client.agents.get(stored.agent_id)
        except Exception:
            # Stale binding — re-register and update storage.
            agent = await client.agents.register(
                public_key=stored.public_key,
                runtime=runtime,
                principal_id=principal_id,
                label=label,
            )
            stored = StoredKey(
                private_key=stored.private_key,
                public_key=stored.public_key,
                created_at=stored.created_at,
                agent_id=agent.agent_id,
                label=label,
            )
            await store.put(name, stored)
    else:
        agent = await client.agents.register(
            public_key=stored.public_key,
            runtime=runtime,
            principal_id=principal_id,
            label=label,
        )
        stored = StoredKey(
            private_key=stored.private_key,
            public_key=stored.public_key,
            created_at=stored.created_at,
            agent_id=agent.agent_id,
            label=label,
        )
        await store.put(name, stored)

    # Step 3 — mint policy.
    expiry = policy_expires_at or (datetime.now(timezone.utc) + DEFAULT_POLICY_TTL)
    policy = await client.policies.create(
        agent.agent_id,
        scopes=scopes or DEFAULT_SCOPES,
        expires_at=expiry,
        label=f"{label}-policy",
    )

    # Step 4 — pre-bound signer closure.
    private_key = stored.private_key
    agent_id = agent.agent_id
    policy_id = policy.policy_id

    async def _sign(**kwargs: Any) -> str:
        ctx = SignContext(**kwargs)
        return sign_agent_token(private_key, agent_id, policy_id, ctx)

    return QuickstartBundle(
        client=client,
        agent=agent,
        policy=policy,
        sign=_sign,
        runtime=detected,
        storage=store,
        key_name=name,
    )


__all__ = ["QuickstartBundle", "quickstart"]

File: packages/sdk-ts/src/quickstart.ts
// Aegis.quickstart() — Round 25 Lane A.
//
// The one-call onboarding helper. Takes an API key + label, returns
// everything a junior developer needs for their first verify:
//
//   const { aegis, agent, policy, sign } = await Aegis.quickstart({
//     apiKey: process.env.AEGIS_API_KEY!,
//     label: 'my-first-agent',
//   });
//   const token = await sign({ action: 'commerce.purchase', amount: 100 });
//   const result = await aegis.verify(token);
//
// What it does:
//
//   1. Reads or generates an Ed25519 keypair via the provided KeyStorage
//      (default: filesystem on Node, IndexedDB on browser, memory on edge).
//   2. Registers the agent with AEGIS (idempotent — if the stored key
//      already has an `agentId`, reuse it).
//   3. Mints a default policy (TTL 24h, action allow-list = `quickstart.*`).
//   4. Returns a `sign(ctx)` closure that pre-binds the agentId + policyId
//      so the developer can sign per-request without passing them.
//
// Junior-grade: every parameter has a sensible default. The only thing
// the caller MUST supply is the API key (or set AEGIS_API_KEY in env).

import { Aegis } from './index.js';
import { generateKeypair, signAgentToken } from './crypto.js';
import type { AgentRecord, PolicyRecord, PolicyScope, SignContext } from './types.js';
import { defaultKeyStorage, type KeyStorage, type StoredKey } from './key-storage.js';
import { detectRuntime, type AegisRuntime } from './runtime.js';

export interface QuickstartOptions {
  /** Required. AEGIS API key. Falls back to `AEGIS_API_KEY` env var. */
  apiKey?: string;
  /** Optional. Defaults to "aegis-quickstart". */
  label?: string;
  /** Optional. Defaults to "ANTHROPIC". */
  runtime?: 'OPENAI' | 'ANTHROPIC' | 'GOOGLE' | 'HUGGINGFACE' | 'CUSTOM';
  /** Optional. Pluggable storage; defaults to runtime-appropriate adapter. */
  storage?: KeyStorage;
  /** Optional key name within storage. Defaults to `label`. */
  keyName?: string;
  /**
   * Optional scope override. Defaults to a permissive `quickstart.*` policy
   * with $100 USD per-transaction cap and a 24-hour expiry — safe for
   * learning, NOT safe for production (kept loose so juniors don't hit
   * SCOPE_NOT_GRANTED on their first verify).
   */
  scopes?: PolicyScope[];
  /** Optional policy expiry. Defaults to 24h from now. */
  policyExpiresAt?: Date;
  /** Optional override of the AEGIS base URL. Defaults to env / public. */
  baseUrl?: string;
}

export interface QuickstartBundle {
  /** The configured Aegis client — pass to relying-party code. */
  aegis: Aegis;
  /** The registered agent (or the one re-bound from stored key). */
  agent: AgentRecord;
  /** The freshly-minted policy. */
  policy: PolicyRecord;
  /** Pre-bound signer — caller passes `SignContext` only. */
  sign: (ctx: SignContext) => Promise<string>;
  /** Detected runtime for telemetry / debugging. */
  runtime: AegisRuntime;
  /** The storage adapter that held the keypair. */
  storage: KeyStorage;
  /** Storage key name (in case caller wants to rebind later). */
  keyName: string;
}

const DEFAULT_LABEL = 'aegis-quickstart';
const DEFAULT_RUNTIME = 'ANTHROPIC' as const;
const DEFAULT_POLICY_TTL_MS = 24 * 60 * 60 * 1000;
const DEFAULT_SCOPES: PolicyScope[] = [
  {
    category: 'commerce',
    spendLimit: { currency: 'USD', maxPerTransaction: 100, maxPerDay: 500 },
  },
];

/**
 * Resolve `AEGIS_API_KEY` from the explicit option, falling back to env.
 * Throws with the catalog-aligned `next` message when neither is set.
 */
function resolveApiKey(explicit: string | undefined): string {
  if (explicit && explicit.length > 0) return explicit;
  // type-rationale: globalThis.process is Node-shaped; we narrow defensively.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const proc = (globalThis as any).process;
  const fromEnv: unknown = proc?.env?.AEGIS_API_KEY;
  if (typeof fromEnv === 'string' && fromEnv.length > 0) return fromEnv;
  throw new Error(
    'Aegis.quickstart: apiKey not provided. ' +
      'Set AEGIS_API_KEY in your environment, or pass apiKey to the SDK constructor ' +
      '(https://docs.aegislabs.io/errors/auth_required)',
  );
}

export async function quickstart(opts: QuickstartOptions = {}): Promise<QuickstartBundle> {
  const apiKey = resolveApiKey(opts.apiKey);
  const label = opts.label ?? DEFAULT_LABEL;
  const keyName = opts.keyName ?? label;
  const storage = opts.storage ?? defaultKeyStorage();
  const runtime = opts.runtime ?? DEFAULT_RUNTIME;
  const detected = detectRuntime();

  const aegis = new Aegis({
    apiKey,
    ...(opts.baseUrl ? { baseUrl: opts.baseUrl } : {}),
  });

  // Step 1 — load or generate the keypair.
  let stored = await storage.get(keyName);
  if (!stored) {
    const kp = await generateKeypair();
    stored = {
      privateKey: kp.privateKey,
      publicKey: kp.publicKey,
      createdAt: new Date().toISOString(),
      label,
    };
    await storage.put(keyName, stored);
  }

  // Step 2 — register the agent (or reuse the stored binding).
  let agent: AgentRecord;
  if (stored.agentId) {
    try {
      agent = await aegis.agents.get(stored.agentId);
    } catch {
      // Stored binding is stale (revoked, deleted, principal changed). Fall
      // through to re-register. We do not silently swallow other errors —
      // the get() call will re-throw transport failures.
      agent = await aegis.agents.register({
        publicKey: stored.publicKey,
        runtime,
        label,
      });
      await storage.put(keyName, { ...stored, agentId: agent.agentId });
    }
  } else {
    agent = await aegis.agents.register({
      publicKey: stored.publicKey,
      runtime,
      label,
    });
    const updated: StoredKey = { ...stored, agentId: agent.agentId };
    await storage.put(keyName, updated);
    stored = updated;
  }

  // Step 3 — mint a default policy.
  const expiresAt = opts.policyExpiresAt ?? new Date(Date.now() + DEFAULT_POLICY_TTL_MS);
  const policy = await aegis.policies.create(agent.agentId, {
    label: `${label}-policy`,
    scopes: opts.scopes ?? DEFAULT_SCOPES,
    expiresAt,
  });

  // Step 4 — return the bundle with a pre-bound signer closure.
  const privateKey = stored.privateKey;
  const agentId = agent.agentId;
  const policyId = policy.policyId;
  const sign = (ctx: SignContext): Promise<string> =>
    signAgentToken(privateKey, agentId, policyId, ctx);

  return {
    aegis,
    agent,
    policy,
    sign,
    runtime: detected,
    storage,
    keyName,
  };
}
